cub/test/catch2_test_thread_scan_exclusive_partial.cu

```diff
@@ -0,0 +1,424 @@
+
+using narrow_precision_type_list = c2h::type_list<
+#if TEST_HALF_T()
+  type_pair<half_t>,
+#endif // TEST_HALF_T()
+#if TEST_BF_T()
+  type_pair<bfloat16_t>
+#endif // TEST_BF_T()
+  >;
+
+using integral_type_list =
+  c2h::type_list<type_pair<cuda::std::uint8_t, cuda::std::int32_t>,
+                 type_pair<cuda::std::int16_t>,
+                 type_pair<cuda::std::uint32_t>,
+                 type_pair<cuda::std::int64_t>>;
+
+using fp_type_list = c2h::type_list<type_pair<float>, type_pair<double>>;
+
+using cub_operator_integral_list =
+  c2h::type_list<cuda::std::plus<>, cuda::std::multiplies<>, cuda::std::bit_and<>, cuda::minimum<>>;
+
+using cub_operator_fp_list = c2h::type_list<cuda::std::plus<>, cuda::std::multiplies<>, cuda::maximum<>>;
+
+static_assert(max_size > 4);
+using items_per_thread_list = c2h::enum_type_list<int, 1, 3, max_size - 1, max_size>;
+
+/***********************************************************************************************************************
+ * Test cases
+ **********************************************************************************************************************/
+
+C2H_TEST("ThreadScanExclusive Integral Type Tests",
+         "[scan][thread]",
+         integral_type_list,
+         cub_operator_integral_list,
+         items_per_thread_list)
+{
+  using params                     = params_t<TestType>;
+  using value_t                    = typename params::item_t;
+  using output_t                   = typename params::output_t;
+  using op_t                       = c2h::get<1, TestType>;
+  using accum_t                    = cuda::std::__accumulator_t<op_t, value_t>;
+  constexpr int num_items          = c2h::get<2, TestType>::value;
+  using dist_param                 = dist_interval<value_t, op_t, num_items, accum_t, output_t>;
+  using filler_dist_param          = dist_interval<accum_t, op_t, num_items, accum_t, output_t>;
+  constexpr auto scan_op           = op_t{};
+  constexpr auto operator_identity = cub_operator_to_identity<accum_t, op_t>::value();
+  const int valid_items            = GENERATE_COPY(
+    take(1, random(2, cuda::std::max(2, num_items - 1))),
+    take(1, random(num_items + 2, cuda::std::numeric_limits<int>::max())),
+    values({1, num_items, num_items + 1}));
+  const int bounded_valid_items = std::min(valid_items, num_items);
+  const accum_t prefix          = GENERATE_COPY(take(1, random(dist_param::min(), dist_param::max())));
+  const bool apply_prefix       = GENERATE(true, false);
+  const accum_t filler          = GENERATE(take(1, random(filler_dist_param::min(), filler_dist_param::max())));
+  CAPTURE(c2h::type_name<value_t>(), num_items, c2h::type_name<op_t>(), valid_items, prefix, apply_prefix);
+
+  c2h::device_vector<value_t> d_in(num_items);
+  c2h::device_vector<output_t> d_out(num_items);
+  c2h::gen(C2H_SEED(num_seeds), d_in, dist_param::min(), dist_param::max());
```

Comment:
Suggestion: use `thrust::no_init` to skip initializing `d_in` since it's overwritten by random numbers anyway. Similary, if `d_out` is overwritten entirely, you can also skip init, since then we can detect partially written data using compute-sanitizer.

Applies to more places.

Reply:
Didn't know that feature yet, very cool!